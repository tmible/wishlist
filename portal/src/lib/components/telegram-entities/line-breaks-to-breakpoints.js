/** @typedef {ReturnType<typeof RegExp.prototype.exec>} Match */
/** @typedef {import('./parser.svelte').Breakpoint} Breakpoint */

/**
 * Отображение переводов строк в объекты, представляющие точки
 * изменения разметки в тексте для каждого перевода строки
 * @function flatMapLineBreaks
 * @param {Match[]} lineBreaks Объекты нахождения переводов строк с помощью регулярного выражения
 * @param {number} textLength Длина всего текста
 * @returns {Breakpoint[]} Точки изменения разметки в тексте
 */
const flatMapLineBreaks = (lineBreaks, textLength) => lineBreaks.flatMap(({ index }, i) => [{
  offset: index,
  entity: {
    type: 'paragraph',
    length: index - (i > 0 ? lineBreaks[i - 1].index + 1 : 0),
  },
  type: 'closing',
}, {
  offset: index,
  entity: {
    type: 'paragraph',
    length: 1,
  },
  type: 'opening',
}, {
  offset: index + 1,
  entity: {
    type: 'paragraph',
    length: 1,
  },
  type: 'closing',
}, {
  offset: index + 1,
  entity: {
    type: 'paragraph',
    length: (
      i < lineBreaks.length - 1 ? lineBreaks[i + 1].index + 1 : textLength
    ) - (index + 1),
  },
  type: 'opening',
}]);

/**
 * Отображение переводов строк в объекты, представляющие точки изменения
 * разметки в тексте. В частности начала и концы абзацев
 * @function lineBreaksToBreakpoints
 * @param {Match[]} lineBreaks Объекты нахождения переводов строк с помощью регулярного выражения
 * @param {number} textLength Длина всего текста
 * @returns {Breakpoint[]} Точки изменения разметки в тексте
 */
export const lineBreaksToBreakpoints = (lineBreaks, textLength) => [
  ...(
    textLength > 0 ?
      [{
        offset: 0,
        entity: {
          type: 'paragraph',
          length: lineBreaks.length > 0 ? lineBreaks[0].index + 1 : textLength,
        },
        type: 'opening',
      }] :
      []
  ),
  ...flatMapLineBreaks(lineBreaks, textLength),
  ...(
    textLength > 0 ?
      [{
        offset: textLength,
        entity: {
          type: 'paragraph',
          length: textLength - (lineBreaks.length > 0 ? lineBreaks.at(-1).index + 1 : 0),
        },
        type: 'closing',
      }] :
      []
  ),
];
