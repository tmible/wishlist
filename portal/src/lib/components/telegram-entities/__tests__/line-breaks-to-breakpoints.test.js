import { describe, expect, it } from 'vitest';
import { lineBreaksToBreakpoints } from '../line-breaks-to-breakpoints.js';

describe('lineBreaksToBreakpoints', () => {
  let text;

  it('should convert line breaks to breakpoints for single line text', () => {
    text = 'single line text';
    expect(
      lineBreaksToBreakpoints(Array.from(text.matchAll(/(?!^)\n(?!$)/g)), text.length),
    ).toMatchSnapshot();
  });

  it('should convert line breaks to breakpoints for text with several lines', () => {
    text = '\nseveral\nlines\ntext\n';
    expect(
      lineBreaksToBreakpoints(Array.from(text.matchAll(/(?!^)\n(?!$)/g)), text.length),
    ).toMatchSnapshot();
  });
});
