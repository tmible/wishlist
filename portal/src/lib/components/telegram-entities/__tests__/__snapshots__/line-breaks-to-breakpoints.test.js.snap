// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lineBreaksToBreakpoints > should convert line breaks to breakpoints for single line text 1`] = `
[
  {
    "entity": {
      "length": 16,
      "type": "paragraph",
    },
    "offset": 0,
    "type": "opening",
  },
  {
    "entity": {
      "length": 16,
      "type": "paragraph",
    },
    "offset": 16,
    "type": "closing",
  },
]
`;

exports[`lineBreaksToBreakpoints > should convert line breaks to breakpoints for text with several lines 1`] = `
[
  {
    "entity": {
      "length": 9,
      "type": "paragraph",
    },
    "offset": 0,
    "type": "opening",
  },
  {
    "entity": {
      "length": 8,
      "type": "paragraph",
    },
    "offset": 8,
    "type": "closing",
  },
  {
    "entity": {
      "length": 1,
      "type": "paragraph",
    },
    "offset": 8,
    "type": "opening",
  },
  {
    "entity": {
      "length": 1,
      "type": "paragraph",
    },
    "offset": 9,
    "type": "closing",
  },
  {
    "entity": {
      "length": 6,
      "type": "paragraph",
    },
    "offset": 9,
    "type": "opening",
  },
  {
    "entity": {
      "length": 5,
      "type": "paragraph",
    },
    "offset": 14,
    "type": "closing",
  },
  {
    "entity": {
      "length": 1,
      "type": "paragraph",
    },
    "offset": 14,
    "type": "opening",
  },
  {
    "entity": {
      "length": 1,
      "type": "paragraph",
    },
    "offset": 15,
    "type": "closing",
  },
  {
    "entity": {
      "length": 5,
      "type": "paragraph",
    },
    "offset": 15,
    "type": "opening",
  },
  {
    "entity": {
      "length": 5,
      "type": "paragraph",
    },
    "offset": 20,
    "type": "closing",
  },
]
`;
